File: MinimumPathSum.java
package leetcode61_70;

/**
 * Given a m x n grid filled with non-negative numbers, find a path
 * from top left to bottom right which minimizes the sum of all numbers along its path.
 Note: You can only move either down or right at any point in time.
 * Created by DCLab on 12/17/2015.
 */
public class MinimumPathSum {

    public static void main(String[] args) {
        int[][] grid = {{0,0,0},
                                {0,1,0},
                                {0,0,0}};
//        int[][] grid = {{1}};
        System.out.println("*****RESULT*****");
        System.out.println(new MinimumPathSum().minPathSum(grid));
    }


    /**
     * 动态规划：res[i][j]=grid[i][j] + Min(res[i-1][j], res[i][j-1])
     * @param grid
     * @return
     */
    public int minPathSum(int[][] grid) {
        int m = grid.length;
        int n = grid[0].length;
        int[][] res = new int[m][n];
        int i=0, j=0;
        res[0][0] = grid[0][0];
        for (i=1; i<m; i++){
            res[i][0] = grid[i][0] + res[i-1][0];
        }
        for (j=1; j<n; j++){
            res[0][j] = grid[0][j] + res[0][j-1];
        }
        for (i=1; i<m; i++){
            for (j=1; j<n; j++){
                res[i][j] = grid[i][j] + Math.min(res[i-1][j], res[i][j-1]);
            }
        }
        return res[i-1][j-1];
    }

}

File: C++/MinimumPathSum.cpp

/**
 * Given a m x n grid filled with non-negative numbers, find a path
 * from top left to bottom right which minimizes the sum of all numbers along its path.
 Note: You can only move either down or right at any point in time.
 * Created by DCLab on 12/17/2015.
 */
class Solution {
public:
    int minPathSum(vector<vector<int>>& grid) {
        int m = grid.size(), n = grid[0].size();
        vector<vector<int>> dp(m, vector<int>(n, 0));
        dp[0][0] = grid[0][0];
        for (int i = 1; i < m; ++i) dp[i][0] = grid[i][0] + dp[i-1][0];
        for (int j = 1; j < n; ++j) dp[0][j] = grid[0][j] + dp[0][j-1];
        for (int i = 1; i < m; ++i) {
            for (int j = 1; j < n; ++j) {
                dp[i][j] = grid[i][j] + min(dp[i-1][j], dp[i][j-1]);
            }
        }
        return dp[m-1][n-1];
    }
};
//public class MinimumPathSum {
//
//    //动态规划：res[i][j]=grid[i][j] + Min(res[i-1][j], res[i][j-1])
//    public int minPathSum(int[][] grid) {
//        int m = grid.length, n = grid[0].length;
//        int[][] dp = new int[m][n];
//        dp[0][0] = grid[0][0];
//        for(int i=1; i<m; i++) dp[i][0] = grid[i][0]+dp[i-1][0];
//        for(int j=1; j<n; j++) dp[0][j] = grid[0][j]+dp[0][j-1];
//        for(int i=1; i<m; i++){
//            for(int j=1; j<n; j++){
//                dp[i][j] = Math.min(dp[i-1][j], dp[i][j-1])+grid[i][j];
//            }
//        }
//        return dp[m-1][n-1];
//    }
//
//}
